src/v/kafka/protocol/tests/batch_reader_test.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/protocol/batch_consumer.h"
+#include "kafka/protocol/batch_reader.h"
+#include "redpanda/tests/fixture.h"
+#include "storage/tests/utils/random_batch.h"
+
+SEASTAR_THREAD_TEST_CASE(batch_reader_last_offset) {
+    constexpr model::offset init_offset{0};
+
+    // Create some batches
+    auto input = storage::test::make_random_batches(init_offset, 40);
+    BOOST_REQUIRE(!input.empty());
+    const auto expected_last_offset = input.back().last_offset();
+
+    // Serialise the batches
+    auto mem_res = model::make_memory_record_batch_reader(std::move(input))
+                     .consume(
+                       kafka::kafka_batch_serializer{}, model::no_timeout)
+                     .get();
+
+    // Functionality under test
+    auto crs = kafka::batch_reader(std::move(mem_res.data));
+
+    BOOST_REQUIRE_EQUAL(crs.last_offset(), expected_last_offset);
+}
+
+SEASTAR_THREAD_TEST_CASE(consumer_records_consume_record_batch) {
+    constexpr model::offset init_offset{0};
+
+    // Create some batches
+    auto input = storage::test::make_random_batches(init_offset, 40);
+    BOOST_REQUIRE(!input.empty());
+    const auto expected_last_offset = input.back().last_offset();
+
+    // Serialise the batches
+    auto mem_res = model::make_memory_record_batch_reader(std::move(input))
+                     .consume(
+                       kafka::kafka_batch_serializer{}, model::no_timeout)
+                     .get();
+
+    // Functionality under test
+    auto crs = kafka::batch_reader(std::move(mem_res.data));
+
+    model::offset last_offset{};
+    while (!crs.empty()) {
```

Comment:
Maybe check that the `crc` is not empty at the beginning?

Reply:
It really should have 40 batches in it to start with. If it is empty, the latter test for last_offset should definitely fail.